File: patterns/2020_09_16/source/jsonUsecase.py
from enum import Enum
from dataclasses import dataclass, field, asdict, fields
from typing import List
import json


@dataclass
class Car:
    makeModel: str = "宝马"
    makeYear: int = 2020
    color: str = "black"
    modelType: str = "X7"


@dataclass
class Employee:
    id: int = 1
    name: str = "长不胖的Garfield"
    level: str = "mid-level"
    languages: List[str] = field(default_factory=lambda:  [
                                 "C++", "Python", "Go"])
    car: Car = Car()


def from_dict(klass, d):
    try:
        fieldtypes = {f.name: f.type for f in fields(klass)}
        return klass(**{f: from_dict(fieldtypes[f], d[f]) for f in d})
    except:
        return d


if __name__ == "__main__":
    me = Employee()

    jsonBuffer = json.dumps(asdict(me), ensure_ascii=False, indent=4)
    print(jsonBuffer)

    result = from_dict(Employee, json.loads(jsonBuffer))
    print(result)

File: patterns/2020_09_16/source/jsonUsecase.cpp
#include <nlohmann/json.hpp>
#include <string>
#include <vector>
#include <iostream>

using json = nlohmann::json;

struct Car
{
    std::string makeModel = u8"宝马";
    int makeYear = 2020;
    std::string color = "black";
    std::string modelType = "X7";
};

enum class SkillLevel
{
    junior,
    midlevel,
    senior,
};

struct Employee
{
    int id = 1;
    std::string name = u8"长不胖的Garfield";
    SkillLevel level = SkillLevel::midlevel;
    std::vector<std::string> languages = {"C++", "Python", "Go"};
    Car car;
};

NLOHMANN_JSON_SERIALIZE_ENUM(SkillLevel, {{SkillLevel::junior, "junior"},
                                          {SkillLevel::midlevel, "mid-level"},
                                          {SkillLevel::senior, "senior"}})

void from_json(json const &j, Car &obj)
{
    j.at("makeModel").get_to(obj.makeModel);
    j.at("makeYear").get_to(obj.makeYear);
    j.at("color").get_to(obj.color);
    j.at("modelType").get_to(obj.modelType);
}

void to_json(json &j, Car const &obj)
{
    j = json{
        {"makeModel", obj.makeModel},
        {"makeYear", obj.makeYear},
        {"color", obj.color},
        {"modelType", obj.modelType}};
}

void from_json(json const &j, Employee &obj)
{
    j.at("id").get_to(obj.id);
    j.at("name").get_to(obj.name);
    j.at("level").get_to(obj.level);
    j.at("languages").get_to(obj.languages);
    j.at("car").get_to(obj.car);
}

void to_json(json &j, Employee const &obj)
{
    j = json{
        {"id", obj.id},
        {"name", obj.name},
        {"level", obj.level},
        {"languages", obj.languages},
        {"car", obj.car}};
}

int main(int argc, char **argv)
{
    Employee me;
    json j1;
    to_json(j1, me);

    auto jsonBuffer = j1.dump(4);
    std::cout << jsonBuffer << "\n";

    Employee result;
    auto j2 = json::parse(jsonBuffer);
    from_json(j2, result);
    std::cout << "Employee:" << result.name << "\n";

    return 0;
}